### ISSUE TITLE
Failed to port a highly recursive algorithm

### ISSUE THREAD

**Issue by pjb7687** [emscripten-core/emscripten#4985]
Hello,

Recently I am trying to port a cover tree library in order to calculate k-nearest neighbors:

http://hunch.net/~jl/projects/cover_tree/cover_tree.html

It compiles very well with Emscripten, but with a large dataset it fails to run, because it tries to make a lot of recursive function calls - and javascript console says 'maximum stack size exceeded'.

Is there any way to avoid this? For example, a command line argument to configure Emscripten not to use actual function calls but to use goto/loop statement, etc.

Best regards,
Jeongbin

**Comment by kripken** [comment]
Which browser is that in? If just one, it could be a bug, and worth filing for them.

If it's all of them, then one option is to refactor the code into a loop as you said. There isn't an automatic system for that, but it's an interesting idea to add one.

Another option, if performance isn't crucial for you, is to build with the emterpreter `-s EMTERPRETIFY=1`. That emulates the code, so it's slow, but it should use far less native stack.

**Comment by pjb7687** [comment]
I tested two browsers: Google Chrome, Firefox

But both fails due to the same reason. I tested the algorithm with several datasets, this only occurs with points > 1000 sample, probably it could be due to so many recursions used in the algorithm.

But it is so weird that it exceeds maximum recursion limit of Firefox, which would be at least 281810 times (this value is for Firefox 42), I hardly believe that this algorithm generates function calls more than this number with just 1000 points of sample... If it is true that it really generates such a lot of recursions, then I would be so happy if Emscripten can deal with such an algorithm in the future.

Anyway thank you for your suggestion for the command line argument (EMTERPRETIFY).
